doc/design/draft/subscription.md

```diff
@@ -0,0 +1,180 @@
+| P2        | 22        | G      |
+
+#### Subscription with _HistorySize_ 20
+
+The subscriber cache contains the ten samples ordered after _DeliveryTime_.
+It contains only ten since there are no more samples available on publisher side.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+| P2        | 24        | H      |
+| P2        | 22        | G      |
+| P1        | 21        | C      |
+| P2        | 20        | F      |
+| P1        | 11        | B      |
+| P1        | 1         | A      |
+
+### Subscribing with concurrent send
+
+Let's assume publisher P1 would like to deliver the sample K at timepoint
+52 while the subscriber requested subsription. Furthermore, the subscriber
+requests always a _HistorySize_ of 2.
+
+All the listed cases should behave identical if there is more then one
+concurrent publish in progress.
+
+#### Publisher before _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+
+The sample K will be delivered after successful subscription.
+
+#### Publisher after _DeliveryTime_
+
+The subscriber cache contains two samples.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P1        | 51        | K      |
+| P2        | 42        | J      |
+
+The sample K will not be delivered by the publisher since the _DeliveryTime_
+is defined as exactly the timepoint after the sample is placed into the ChunkHistory
+and the subscriber cache is based on that ChunkHistory.
+
+### Subscribing After Publisher Stops Offering
+
+In this case we assume that publisher P1 has stopped offering the service
+before the subscription was requested.
+
+#### Subscription with _HistorySize_ 4
+
+The subscriber cache contains the four samples ordered after _DeliveryTime_.
+
+| Publisher | Timepoint | Sample |
+|:---------:|:---------:|:------:|
+| P2        | 42        | J      |
+| P1        | 41        | E      |
+| P2        | 40        | I      |
+| P1        | 31        | D      |
+
+
+
+## Design
+
+### Idea 1: One Ring Buffer Which Contains The Samples
+
+- All publishers are sharing one random access lock-free ring-buffer.
+- The ring-buffer is filled with shared-chunks which can be copied in a threadsafe
+    and lock-free manner.
+- A publisher delivers a sample by writing that sample into the shared ring-buffer.
+- Every subscriber owns an index to their own read position of the ring-buffer.
+- On subscription the subscriber will acquire a const reference to the ring-buffer.
+- When subscriber acquires a sample with take, the shared-chunk is copied from
+    the index position and the index is incremented.
+- When the publisher delivers a sample it will verify all subscriber indices.
```

Comment:
I would see the responsibility here at subscriber side to know which data it wants to extract from the cache (again the overflow behavior and other issues).
Ideally each data item gets a sequence number when it is placed in the cache (this also generats an ordering for all subscribers then!) and the subscriber can verify whether it has consecutive data. However, losing data is inevitable with any restricted capacity buffer approach. That is why I consider storing read indices for individual subscribers meaningless.

Reply:
Addendum: the point of caching is to provide samples to late joining subscribers. Hence they are not known before they subscribe and we cannot keep any index specifically for them for this reason. Once they join, the get the requested samples from the cache (if any). After this we could keep an index for a subscriber, but then we would use this structure instead of the MPMC queue, which would be vastly more difficult since it would require non-destructive reading for multiple readers. I do not know how to solve this and I presume it is harder than the MPMC queue (it is more general).